plugins/hls-notes-plugin/src/Ide/Plugin/Notes/Internal.hs

```diff
@@ -0,0 +1,89 @@
+module Ide.Plugin.Notes.Internal where
+
+import           Control.Lens               ((^.))
+import           Control.Monad.IO.Class     (liftIO)
+import qualified Data.Array                 as A
+import           Data.Bifunctor             (first)
+import qualified Data.ByteString.Char8      as BS
+import           Data.Maybe                 (listToMaybe, mapMaybe)
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as TE
+import qualified Data.Text.Encoding.Error   as TE
+import           Development.IDE            (IdeState, Location (..),
+                                             Range (..))
+import           Development.IDE.Core.Rules (getSourceFileSource, runAction)
+import           Ide.PluginUtils            (getNormalizedFilePath,
+                                             pluginResponse, throwPluginError)
+import           Ide.Types                  (PluginMethodHandler)
+import           Language.LSP.Types         (Method (TextDocumentDefinition),
+                                             Position (..), UInt,
+                                             type (|?) (..))
+import qualified Language.LSP.Types.Lens    as L
+import           Prelude                    hiding (lines, mod)
+import           Text.Regex.TDFA            (CompOption (..),
+                                             RegexOptions (defaultExecOpt),
+                                             defaultCompOpt, makeRegexOpts,
+                                             matchAllText)
+import           Text.Regex.TDFA.Text       (Regex)
+
+jumpToNote :: PluginMethodHandler IdeState TextDocumentDefinition
+jumpToNote state _ param = pluginResponse $ do
+    let uriOrig = param ^. (L.textDocument . L.uri)
+    nfp <- getNormalizedFilePath uriOrig
+
+    file <- liftIO $ runAction "Notes.fileContents" state (getSourceFileSource nfp)
+    let lines = BS.split '\n' file
```

Comment:
There's a bunch of "linear in the length of the file" stuff in here. That seems unavoidable, since you're searching files for notes, but we _definitely_ don't want to do this in a handler like this otherwise things will get quite bad on big files.

The thing to do is to define a rule that finds all the note definitions in the file and stores them in a structure that then permits quick lookup. That way we can run and cache it separately from the handler, and the handler can be fast. You'll also definitely need something like this to get definitions from other files. 

You can find lots of examples of this pattern in other plugins.